sql/src/main/java/io/crate/metadata/functions/params/Param.java

```diff
@@ -0,0 +1,270 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.metadata.functions.params;
+
+import com.google.common.base.Preconditions;
+import io.crate.analyze.symbol.FuncArg;
+import io.crate.exceptions.ConversionException;
+import io.crate.types.ArrayType;
+import io.crate.types.BooleanType;
+import io.crate.types.CollectionType;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import io.crate.types.IntegerType;
+import io.crate.types.SetType;
+import io.crate.types.StringType;
+
+import javax.annotation.Nullable;
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Objects;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+/**
+ * A single function parameter as part of a function parameter definition in {@link FuncParams}.
+ */
+public final class Param {
+
+    public static final Param ANY = of();
+    public static final Param NUMERIC = of(DataTypes.NUMERIC_PRIMITIVE_TYPES);
+    public static final Param ANY_ARRAY = of(new ArrayType());
+    public static final Param ANY_SET = of(new SetType());
+    public static final Param INTEGER = of(IntegerType.INSTANCE);
+    public static final Param STRING = of(StringType.INSTANCE);
+    public static final Param BOOLEAN = of(BooleanType.INSTANCE);
+
+    private final SortedSet<DataType> validTypes;
+    private final ThreadLocal<FuncArg> boundType;
+    private final ThreadLocal<Deque<DataType>> multiBoundStack;
+
+    @Nullable
+    private final Param innerType;
+
+    private Param(DataType... validTypes) {
+        this(Collections.emptyList(), validTypes, null);
+    }
+
+    private Param(Collection<DataType> validTypes, DataType[] validTypes2, @Nullable Param innerType) {
+        this.boundType = new ThreadLocal<>();
+        this.validTypes = new TreeSet<>((o1, o2) -> {
+            if (o1.precedes(o2)) {
```

Comment:
👍

Reply:
I saw that you changed this but reverted this, what was the reason? (It's okay, just curious)

Reply:
Forgot to drop you a comment. I removed this because I need the valid types in descending precedence order. When I use the default compare to, I would have to negate the result. Maybe that's still easier to read but I didn't want to fiddle with it too much.

Reply:
This would work:
```java
new Comparator<DataType>() {
        @Override
        public int compare(DataType o1, DataType o2) {
            return o1.precedence().compareTo(o2.precedence());
        }
}.reversed());
```